serverless/lambda/src/main/java/uk/gov/di/lambdas/LogoutHandler.java

```diff
@@ -4,29 +4,164 @@
+                            t -> {
+                                final ClientRegistry clientRegistry;
+                                try {
+                                    clientRegistry =
+                                            dynamoClientService
+                                                    .getClient(t)
+                                                    .orElseThrow(
+                                                            () -> new ClientNotFoundException(t));
+                                } catch (ClientNotFoundException e) {
+                                    throw new RuntimeException(
+                                            format(
+                                                    "Client not found in ClientRegistry for ClientID: %s",
+                                                    t));
+                                }
+                                String logoutURI =
+                                        validateClientRedirectUri(
+                                                queryStringParameters, clientRegistry);
+                                return generateLogoutResponse(URI.create(logoutURI), state);
+                            })
+                    .orElse(generateDefaultLogoutResponse(state));
+        } catch (ParseException e) {
+            throw new RuntimeException();
+        }
+    }
+
+    private boolean isIDTokenSignatureValid(String idTokenHint, String sessionID) {
+        return true;
+    }
+
+    private boolean doesIDTokenExistInSession(String idTokenHint, Session session) {
+        for (Map.Entry<String, ClientSession> t : session.getClientSessions().entrySet()) {
+            boolean idTokenHintExists =
+                    t.getValue().getIdTokenHint() != null
+                            && t.getValue().getIdTokenHint().equals(idTokenHint);
+            if (idTokenHintExists) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String validateClientRedirectUri(
+            Map<String, String> queryStringParameters, ClientRegistry clientRegistry) {
+        String postLogoutRedirectUri = queryStringParameters.get("post_logout_redirect_uri");
+        if (!queryStringParameters.get("post_logout_redirect_uri").isBlank()
+                && clientRegistry.getPostLogoutRedirectUrls().contains(postLogoutRedirectUri)) {
+            return postLogoutRedirectUri;
+        }
+        return configurationService.getDefaultLogoutURI().toString();
+    }
+
+    private APIGatewayProxyResponseEvent generateDefaultLogoutResponse(Optional<String> state) {
+        return generateLogoutResponse(configurationService.getDefaultLogoutURI(), state);
+    }
+
+    private APIGatewayProxyResponseEvent generateLogoutResponse(
+            URI logoutUri, Optional<String> state) {
+        URIBuilder uriBuilder = new URIBuilder(logoutUri);
+        state.map(s -> uriBuilder.addParameter("state", s));
```

Comment:
```suggestion
        state.ifPresent(s -> uriBuilder.addParameter("state", s));
```

We don't care about the return value since we're just mutating the uriBuilder off the optional if it's there

Reply:
Good shout. I have updated.